core/src/queue.rs

```diff
@@ -220,64 +219,51 @@ impl Queue {
         }
     }
 
-    /// Pop a single transaction.
-    ///
-    /// Unlike [`Self::pop`], unsigned transactions are not recorded.
-    #[allow(
-        clippy::expect_used,
-        clippy::unwrap_in_result,
-        clippy::cognitive_complexity
-    )]
-    pub fn pop_without_seen(&self, wsv: &WorldStateView) -> Option<VersionedAcceptedTransaction> {
-        loop {
-            let hash = self.queue.pop()?;
-            let entry = match self.txs.entry(hash) {
-                Entry::Occupied(entry) => entry,
-                // As practice shows this code is not `unreachable!()`.
-                // When transactions are submitted quickly it can be reached.
-                Entry::Vacant(_) => continue,
-            };
-            if self.check_tx(entry.get(), wsv).is_err() {
-                entry.remove_entry();
-                continue;
-            }
-
-            if *entry
-                .get()
-                .check_signature_condition(wsv)
-                .expect("Checked in `check_tx` just above")
-            {
-                return Some(entry.get().clone());
-            }
-        }
-    }
-
-    /// Return the number of transactions in the queue
+    /// Return the number of transactions in the queue.
     pub fn tx_len(&self) -> usize {
         self.txs.len()
     }
 
     /// Gets transactions till they fill whole block or till the end of queue.
     ///
     /// BEWARE: Shouldn't be called in parallel with itself.
-    #[allow(clippy::missing_panics_doc, clippy::unwrap_in_result)]
-    pub fn get_transactions_for_block(
+    #[cfg(test)]
+    fn collect_transactions_for_block(
         &self,
         wsv: &WorldStateView,
     ) -> Vec<VersionedAcceptedTransaction> {
+        let mut transactions = Vec::with_capacity(self.txs_in_block);
+        self.put_transactions_for_block_into(wsv, &mut transactions);
+        transactions
+    }
+
+    /// Put transactions into provided vector until they fill the whole block or there are no more transactions in the queue.
+    ///
+    /// BEWARE: Shouldn't be called in parallel with itself.
+    pub fn put_transactions_for_block_into(
+        &self,
+        wsv: &WorldStateView,
+        transactions: &mut Vec<VersionedAcceptedTransaction>,
+    ) {
+        if transactions.len() >= self.txs_in_block {
+            return;
+        }
+
         let mut seen = Vec::new();
 
-        let out = std::iter::repeat_with(|| self.pop(&mut seen, wsv))
-            .take_while(Option::is_some)
-            .map(Option::unwrap)
-            .take(self.txs_in_block)
-            .collect::<Vec<_>>();
+        let transactions_hashes: HashSet<HashOf<VersionedSignedTransaction>> = transactions
+            .iter()
+            .map(VersionedAcceptedTransaction::hash)
+            .collect();
+        let out = std::iter::from_fn(|| self.pop(&mut seen, wsv))
+            .filter(|tx| !transactions_hashes.contains(&tx.hash()))
```

Comment:
On the second look i have discovered that `transactions.insert` is redundant here, because since we store transactions in the queue untill expired or in the blockchain in would be impossible to receive duplicate transaction from the queue.